wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java

```diff
@@ -379,27 +382,51 @@ public void checkout(CheckoutParams params) throws GitException {
       if (startPoint != null && trackBranch != null) {
         throw new GitException("Start point and track branch can not be used together.");
       }
-      if (params.isCreateNew() && name == null) {
+      if (params.isCreateNew() && branchName == null) {
         throw new GitException("Branch name must be set when createNew equals to true.");
       }
       if (startPoint != null) {
         checkoutCommand.setStartPoint(startPoint);
       }
       if (params.isCreateNew()) {
         checkoutCommand.setCreateBranch(true);
-        checkoutCommand.setName(name);
-      } else if (name != null) {
-        checkoutCommand.setName(name);
+        checkoutCommand.setName(branchName);
+      } else if (branchName != null) {
+        checkoutCommand.setName(branchName);
         List<String> localBranches =
             branchList(LIST_LOCAL)
                 .stream()
                 .map(Branch::getDisplayName)
                 .collect(Collectors.toList());
-        if (!localBranches.contains(name)) {
+        if (!localBranches.contains(branchName)) {
+          List<Branch> branches = branchList(LIST_REMOTE);
+          Set<String> remotes =
+              repository.getConfig().getSubsections(ConfigConstants.CONFIG_KEY_REMOTE);
+          long count =
+              remotes
+                  .stream()
+                  .filter(
+                      remote ->
+                          branches
+                              .stream()
+                              .anyMatch(
+                                  branch ->
+                                      ("refs/remotes/" + remote + "/" + branchName)
+                                          .equals(branch.getName())))
+                  .count();
+          if (count > 1 || count == 0) {
+            throw new GitException(String.format(ERROR_CHECKOUT_BRANCH_NAME_NOT_FOUND, branchName));
+          }
+
           Optional<Branch> remoteBranch =
-              branchList(LIST_REMOTE)
+              branches
                   .stream()
-                  .filter(branch -> branch.getName().contains(name))
+                  .filter(
+                      branch ->
+                          remotes
+                              .stream()
+                              .anyMatch(
+                                  remote -> branch.getName().endsWith(remote + "/" + branchName)))
```

Comment:
I guess there is a way to use the only stream iteration.

Reply:
done